# Tangelo hands-on: Phase Estimation

For more information about quantum phase estimation, you can check out the [wikipedia page](https://en.wikipedia.org/wiki/Quantum_phase_estimation_algorithm). Essentially, by performing a bunch of controlled- time-evolutions we can determine the energy of a state. If that state is not an eigenstate, the measurement at the end will collapse the eigenstate to something closer to the eigenstate with the energy measured.

## Before you jump in

Before tackling this hands-on, we suggest to have a look at:

- the tutorial on the basics of the linq module

You can use the cells below to install or import useful packages. We recommend to install qulacs as some of these circuits could be long.

In [66]:
import numpy as np

If you wish to install anything in your jupyter environment while working on this notebook, including Tangelo, you can go through your terminal and reload the upyter kernel or run `pip` commands directly from this notebook using `!`

In [67]:
# Uncomment and change package name to install it in the current environment
# !pip install tangelo-gc --quiet
# !pip install qulacs --quiet

## Level 1: QPE with exact eigenstate

Let's define a molecule in second quantization, for the sake of exploring the variational framework. Here we define a molecule explicitly, specifying a geometry, charge, spin and basis set, but you could also quickly just import other existing ones in our ["molecule library"](https://github.com/goodchemistryco/Tangelo/blob/main/tangelo/molecule_library.py) if you want to swap between other examples.

In [68]:
from openfermion import get_sparse_operator

from tangelo.toolboxes.operators import QubitOperator

# Generate qubit operator with state 9 having eigenvalue 0.25
qu_op = QubitOperator("X0 X1", 0.125) + QubitOperator("Y1 Y2", 0.125) + QubitOperator("Z2 Z3", 0.125) + QubitOperator("", 0.125)

ham_mat = get_sparse_operator(qu_op).toarray()
eigs, wavefunctions = np.linalg.eigh(ham_mat)

# Eigenvector 9 has eigenvalue 0.25
print(eigs[9])

0.24999999999999992


In [69]:
from tangelo.linq import get_backend

# Choose backend with lsq_first (least significant qubit first) ordering
sim = get_backend("cirq")
print(f"sim has statevector ordering of {sim.backend_info()['statevector_order']}")

sim has statevector ordering of lsq_first


In [70]:
from tangelo.linq.helpers.circuits.statevector import StateVector

# Create the circuit that initialized the 9th
sv = StateVector(wavefunctions[:, 9].copy(), order=sim.backend_info()['statevector_order'])

sv_circuit = sv.initializing_circuit()


In [71]:
# check that state has correct energy
print(f"sv_circuit prepared a state with energy {sim.get_expectation_value(qubit_operator=qu_op, state_prep_circuit=sv_circuit)}")
n_state_qubits = sv_circuit.width
print(f"The state register is composed of {n_state_qubits}")

sv_circuit prepared a state with energy 0.2500000000000001
The state register is composed of 4


Now you can prepare the circuit for phase estimation. Remember this figure.

![Phase Estimation Circuit](PE_Circuit.png "Phase Estimation Circuit")

In [72]:
from tangelo.toolboxes.ansatz_generator.ansatz_utils import trotterize, get_qft_circuit
from tangelo.toolboxes.post_processing.histogram import Histogram
from tangelo.linq import Gate, Circuit

# Reverse order as cirq uses lsq_first
qubit_list = [6, 5, 4]

# State preparation
pe_circuit = sv_circuit + Circuit([Gate("H", q) for q in qubit_list ])

for i, qubit in enumerate(qubit_list):
    # You can play around with how accurate the time-evolution needs to be
    # Use negative time as trotterize uses exp(-iHt) to follow the Schrodinger equation
    pe_circuit += trotterize(qu_op, trotter_order=4, n_trotter_steps=4, time=-2*np.pi*2**i, control=qubit)

# set inverse to true or false
pe_circuit += get_qft_circuit(qubit_list, inverse=True)

freqs, _ = sim.simulate(pe_circuit)

# Remove qubit indices from histogram corresponding to the state qubits i.e. (0, 1, 2, 3)
hist = Histogram(freqs)
hist.remove_qubit_indices(0, 1, 2, 3)

for key, probability in hist.frequencies.items():
      energy = sum(int(k)/2**(i+1) for i, k in enumerate(key))
      print(f"The probability that the statevector energy={energy} is {probability:3.5f}")

The probability that the statevector energy=0.25 is 0.99999
The probability that the statevector energy=0.0 is 0.00000
The probability that the statevector energy=0.125 is 0.00000
The probability that the statevector energy=0.375 is 0.00000
The probability that the statevector energy=0.5 is 0.00000
The probability that the statevector energy=0.625 is 0.00000
The probability that the statevector energy=0.875 is 0.00000


## Approximate example

In [104]:
# Define H2 molecule in minimal basis set
from tangelo import SecondQuantizedMolecule
from tangelo.toolboxes.qubit_mappings.mapping_transform import fermion_to_qubit_mapping
from tangelo.toolboxes.qubit_mappings.statevector_mapping import get_reference_circuit

xyz_H2 = [("H", (0., 0., 0.)), ("H", (0., 0., 0.7414))]
mol_H2_sto3g = SecondQuantizedMolecule(xyz_H2, q=0, spin=0, basis="sto-3g")

mol = mol_H2_sto3g

qu_op = fermion_to_qubit_mapping(mol.fermionic_hamiltonian, mapping="JW", n_spinorbitals=mol.n_active_sos, n_electrons=mol.n_active_electrons, spin=mol.spin, up_then_down=False)

# Cheat to get the ground state eigenvalue to be 0.125mod(1) exact ground state is 1.137270174660903
qu_op += (0.25 + 1.137270174660903)
eigs, statevectors = np.linalg.eigh(get_sparse_operator(qu_op).toarray())

print(eigs[0])
ground_sv = statevectors[:, 0]
print(eigs[13])
first_sv = statevectors[:, 13]

0.24999999999999947
1.8671062929051834


In [125]:
# Hartree-Fock reference state circuit
hf_circuit = get_reference_circuit(n_spinorbitals=mol.n_active_sos, n_electrons=mol.n_active_electrons, mapping="JW", up_then_down=False, spin=mol.spin)

f, sv = sim.simulate(hf_circuit, return_statevector=True)

g_ovlp = np.dot(sv, ground_sv)
f_ovlp = np.dot(sv, first_sv)
print(f"The overlap of the initial state with the exact ground state is {g_ovlp}")
print(f"The overlap of the initial state with the exact first excited state is {f_ovlp}")

The overlap of the initial state with the exact ground state is (0.9936146058054713+0j)
The overlap of the initial state with the exact first excited state is (-0.112827368710068+0j)


Use phase-estimation to do state preparation. Warning, you will probably need very high order time-evolution to get a good result.

In [110]:
# Reverse order as cirq uses lsq_first
qubit_list = [6, 5, 4]

# State preparation
pe_circuit = hf_circuit + Circuit([Gate("H", q) for q in qubit_list ])

for i, qubit in enumerate(qubit_list):
    # You can play around with how accurate the time-evolution needs to be
    # Use negative time as trotterize uses exp(-iHt) to follow the Schrodinger equation
    pe_circuit += trotterize(qu_op, trotter_order=4, n_trotter_steps=12, time=-2*np.pi*2**i, control=qubit)

# set inverse to true or false
pe_circuit += get_qft_circuit(qubit_list, inverse=True)

sim = get_backend("cirq")
freqs, _ = sim.simulate(pe_circuit)

# Remove qubit indices from histogram corresponding to the state qubits i.e. (0, 1, 2, 3)
hist = Histogram(freqs)
hist.remove_qubit_indices(0, 1, 2, 3)

for key, probability in hist.frequencies.items():
      energy = sum(int(k)/2**(i+1) for i, k in enumerate(key))
      print(f"The probability that the statevector energy={energy} is {probability:3.5f}")

The probability that the statevector energy=0.0 is 0.00004
The probability that the statevector energy=0.125 is 0.00001
The probability that the statevector energy=0.25 is 0.98752
The probability that the statevector energy=0.375 is 0.00005
The probability that the statevector energy=0.5 is 0.00001
The probability that the statevector energy=0.625 is 0.00003
The probability that the statevector energy=0.75 is 0.00005
The probability that the statevector energy=0.875 is 0.01228


How good is our final state?

In [117]:
pe_plus_measure = pe_circuit + Circuit([Gate("MEASURE", i) for i in qubit_list])

_, sv_new = sim.simulate(pe_plus_measure, desired_meas_result="010", return_statevector=True)

print(f"The probability of this measurement is {pe_plus_measure.success_probabilities['010']}")

# Shrink vector down to 2**4 size to compare with exact ground state.
sv_new_post = np.reshape(sv_new, (2**4, 2**3))[:, int("010", base=2)]

print(f"The final state overlap with the ground state is {np.abs(np.dot(sv_new_post, ground_sv))}")

The probability of this measurement is 0.9875184986388642
The final state overlap with the ground state is 0.9999727387751328


What if we measured the energy of the first excited state (1.8671 mod(1))= 0.867. The closest measurement is '111' which has energy 1/2+1/4+1/8=0.875. Let's run the calculation again with a desired result of 111.

In [122]:
_, sv_new = sim.simulate(pe_plus_measure, desired_meas_result="111", return_statevector=True)

# Shrink vector down to 2**4 size to compare with exact ground state.
sv_new_post = np.reshape(sv_new, (2**4, 2**3))[:, int("111", base=2)]

print(f"The final state overlap with the first excited state is {np.abs(np.dot(sv_new_post, first_sv))}")

The final state overlap with the first excited state is 0.9998680358098726


Of course, the error on the energy measurement is 0.008 (0.875-0.867) which is not chemically accurate. But the state we prepared is, we just can't easily measure the energy of the state.

All success probabilities are saved in the Circuit so different outcomes can be evaluated without generating the full frequency dictionary. This becomes especially important when many mid-circuit measurements are utilized but only a few are meaningful.

In [129]:
print(f"The success probabilities for the various runs are\n {pe_plus_measure.success_probabilities}")

The success probabilities for the various runs are
 {'010': 0.9875184986388642, '111': 0.012282064140958181}


You can see that these probabilites are very similar to the initial overlaps squared

In [130]:
print(g_ovlp.real**2, f_ovlp.real**2)

0.9872699848699622 0.012730015130037632


Phase estimation is the backbone of many fault-tolerant quantum algorithms. Both as a state preparation method and a component of other algorithms such as HHL and Shor's algorithm.

What will you do with Tangelo ?